interfaces/devdoc/file_requirements.md

```diff
@@ -0,0 +1,116 @@
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_READ_ASYNC_RESULT, file_read_async, FILE_HANDLE, handle, uint32_t, size, uint64_t, position, FILE_READ_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+MOCKABLE_FUNCTION_WITH_RETURNS(, int, file_extend_filesize, FILE_HANDLE, handle, uint64_t, desired_size, bool, has_manage_volume)(0, MU_FAILURE);
+```
+
+## file_create
+
+```c
+MOCKABLE_FUNCTION(, FILE_HANDLE, file_create, EXECUTION_ENGINE_HANDLE, execution_engine, const char*, full_file_name, uint64_t, desired_file_size, bool, has_manage_volume);
+```
+
+`file_create` opens an existing file by the name of `full_file_name` or creates a file of `desired_file_size` if it doesn't exist and returns the file handle.
+
+**SRS_FILE_43_003: [** If a file with name `full_file_name` does not exist, `file_create` shall create a file with that name and size `desired_file_size`. **]**
+
+**SRS_FILE_43_001: [** `file_create` shall open the file named `full_file_name` for asynchronous operations and return its handle. **]**
+
+**SRS_FILE_43_002: [** If `full_file_name` is `NULL` then `file_create` shall fail and return `NULL`. **]**
+
+`file_create` shall call `file_extend_filesize` to extend the file size to `desired_file_size` passing `has_manager_volume_privilege`
+
+**SRS_FILE_43_004: [** If the file is opened successfully, `file_create` shall return `0`. **]**
+
+## file_destroy 
+
+```c
+MOCKABLE_FUNCTION(, void, file_destroy, FILE_HANDLE, handle);
+```
+
+`file_destroy` closes the given file handle.
+
+**SRS_FILE_43_005: [** If `handle` is null, `file_destroy` shall return. **]**
+
+**SRS_FILE_43_006: [** `file_destroy` shall wait for all pending I/O operations to complete. **]**
+
+**SRS_FILE_43_007: [** `file_destroy` shall close the file handle `handle`. **]**
+
+## file_write_async
+
+```c
+MOCKABLE_FUNCTION_WITH_RETURNS(, FILE_WRITE_ASYNC_RESULT, file_write_async, FILE_HANDLE, handle, CONSTBUFFER_HANDLE, source, uint64_t, position, FILE_WRITE_CB, user_callback, void*, user_context)(0, MU_FAILURE);
+```
+
+`file_write_async` issues an asynchronous write request.
+
+**SRS_FILE_43_008: [** `file_write_async` writes `source`'s content to the position offset in the file and calls `user_call_back` passing `user_context` and success.**]**
+
+**SRS_FILE_43_009: [** If `handle` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_010: [** If `source` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_012: [** If `user_callback` is `NULL` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_013: [** If writing `size` bytes at `position` would exceed `desired_file_size` then `file_write_async` shall fail and return `FILE_WRITE_ASYNC_INVALID_ARGS`. **]**
+
+**SRS_FILE_43_014: [** `file_write_async` shall create an event to be used for the asynchronous write operation. **]**

```

Comment:
> shall create an event [](start = 41, length = 22)

Sounds like a win32 specific spec, not for the interface.

Reply:
Yeah I agree that this spec doesn't belong here. Should I add it to win32_reqs though? Because in bsdl_file_ll.c it looks like there isn't really a need for this event.

---
In reply to: [448509531](https://github.com/Azure/azure-c-pal/pull/7#discussion_r448509531) [](ancestors = 448509531)